### ISSUE TITLE
Loss diverging while trying to reproduce product quantization results following fairseq README 

### ISSUE THREAD

**Issue by amandeephy** [facebookresearch/fairseq#4107]
Dear authors,

Thank you for your great work on Fairseq.
When trying to train a language model with product quantization (following the fairseq README : https://github.com/pytorch/fairseq/tree/main/examples/quant_noise#walk-through-the-code) we encounter the below described issue.

## 🐛 Bug

Loss diverging while trying to reproduce product quantization results following fairseq README : https://github.com/pytorch/fairseq/tree/main/examples/quant_noise#walk-through-the-code

### To Reproduce

Command I ran :

```
BASE_DIR=<path_to_base_dir>
MAX_TOKENS=2048

fairseq-train --task language_modeling $BASE_DIR/data-bin/wikitext-103/ \
                    --save-dir $BASE_DIR/tlm_gbw_pq_wikitext-103/checkpoints/ \
                    --adaptive-input --adaptive-input-cutoff 20000,60000 \
                    --adaptive-input-factor 4 \
                    --adaptive-softmax-cutoff 20000,60000 \
                    --adaptive-softmax-dropout 0.2 --adaptive-softmax-factor 4.0 \
                    --tie-adaptive-proj --tie-adaptive-weights \
                    --arch transformer_lm_gbw \
                    --attention-dropout 0.1 --dropout 0.2 --relu-dropout 0.1 \
                    --clip-norm 0.1 --criterion adaptive_loss \
                    --ddp-backend legacy_ddp \
                    --decoder-attention-heads 8 --decoder-embed-dim 1024 \
                    --decoder-ffn-embed-dim 4096 --decoder-input-dim 1024 \
                    --decoder-layers 16 --decoder-normalize-before --decoder-output-dim 1024 \
                    --min-lr 0.0001 --lr-period-updates 270000 --lr-scheduler cosine \
                    --lr-shrink 0.75 --lr 1.0 --t-mult 2.0 \
                    --max-tokens $MAX_TOKENS --tokens-per-sample $MAX_TOKENS \
                    --momentum 0.99 --optimizer nag \
                    --sample-break-mode none --update-freq 3 \
                    --warmup-init-lr 1e-07 --warmup-updates 16000 \
                    --weight-decay 0 --seed 1 --stop-min-lr 1e-09 \
                    --quant-noise-pq 0.05 --quant-noise-pq-block-size 8 \
                    --tensorboard-logdir $BASE_DIR/tlm_gbw_pq_wikitext-103/tensorboard
```
This command is similar to the one provided in the README, with the only change being in the maximum number of tokens (2048 instead of 3072).

### Expected behavior

We were expecting to train a model with product quantization noise, however while training, we see the loss diverge after only a couple of epochs.

Attached below is the loss curve  :
![tensorboard_loss_0_6_smoothing](https://user-images.githubusercontent.com/34727495/148268655-89fa0b1c-f541-4c87-b39d-a6cc62a82281.png)

and the command line output:
![Diverging loss](https://user-images.githubusercontent.com/34727495/148268663-a5a98aa3-ada3-425c-b09b-4b60341dfe0a.png)

### Environment
 - fairseq Version (e.g., 1.0 or main) : '1.0.0a0+7fb3792'
 - PyTorch Version (e.g., 1.0) : '1.10.0'
 - OS (e.g., Linux): Ubuntu 16.04.7 LTS
 - How you installed fairseq (`pip`, source):  source
 - Build command you used (if compiling from source): `pip install --editable ./`
 - Python version: Python 3.8.0
 - CUDA/cuDNN version: CUDA Version: 11.2
 - GPU models and configuration: NVIDIA Tesla V100 (4x)
